Feature implement json logformat

yuzhiquan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
Feature implement json logformat
Part of kubernetes/enhancements#1602 effort

Allow users to change log out of k8s components by passing flag --log-format=json.
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #91490

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

yuzhiquan commented:
Still working on unit test and benchmark.
@serathius PLZ have a look.

serathius commented:
I'm worried about getting approvals for dependencies at this stage. Can you try to avoid adding dependencies in so many directories? would moving `json.go` to subdirectory help?

yuzhiquan commented:
Yes, i worried about this too.
After run 'go get go.uber.org/zap@v1.10.0 ', many go.sum  updated.

yuzhiquan committed:
add json logger format

add unit test case

add benchmark test case

fix staticcheck, and response for review comment

remove unnecessary variable

add test case for non-string field or zap-field, refactor code

update vendor

yuzhiquan commented:
Move json to `staging/src/k8s.io/component-base/logs/json/` subdir, updated less go.sum.

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+		}
+
+		fields = append(fields, zap.Any(keyStr, val))
+		i += 2
```
Move to for loop

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+		}
+
+		// make sure this isn't a mismatched key
+		if i == len(args)-1 {
```
Can we check this before for loop instead of on every iteration?

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+	// fields (since logr has no concept of that), so guess that we need a
+	// little less space.
+	fields := make([]zap.Field, 0, len(args)/2+len(additional))
+	for i := 0; i < len(args); {
```
```suggestion
	for i := 0; i < len(args) - 1; i+=2 {
```
If you apply other suggestion then it should look like this

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+
+// Error write log message to error level
+func (l *zapLogger) Error(err error, msg string, keysAndVals ...interface{}) {
+	if checkedEntry := l.l.Check(zap.ErrorLevel, msg); checkedEntry != nil {
```
Why we need to call `Check` ? In logr we always write a error log.

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+func (l *zapLogger) V(level int) logr.InfoLogger {
+	lvl := zapcore.Level(-1 * level)
+	if l.l.Core().Enabled(lvl) {
+		l.lvl = lvl
```
This would modify and return current logger. For V to work properly we need to copy.

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+
+// V return info logr.Logger with specified level
+func (l *zapLogger) V(level int) logr.InfoLogger {
+	lvl := zapcore.Level(-1 * level)
```
Why use zap Level? I don't think we should map Logr levels to zap levels.

yuzhiquan replied:
We should set zap level in infologger, and followed `l.l.Core().Enabled(lvl)` && [l.l.Check(l.lvl, msg)](https://github.com/kubernetes/kubernetes/blob/287aacb39d8d097791eb1d632b24a2389b14f028/staging/src/k8s.io/component-base/logs/json/json.go#L56) need  zap level.

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+
+// WithValues return logr.Logger with some keys And Values
+func (l *zapLogger) WithValues(keysAndValues ...interface{}) logr.Logger {
+	newLogger := l.l.With(handleFields(l.l, keysAndValues)...)
```
I expect that `WithValues` and `WithName` will be called frequently in future. You use `NewJsonLogger` which looks costly (allocates a lot of new objects on heap)

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+
+// NewJsonLogger creates a new json logr.Logger using the given Zap Logger to log.
+func NewJsonLogger(l *zap.Logger, w zapcore.WriteSyncer) logr.Logger {
+	encoderConfig := zapcore.EncoderConfig{
```
This config never changes, can we move it out of this function?

serathius commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,190 @@
+		EncodeTime: zapcore.EpochMillisTimeEncoder,
+	}
+
+	var syncer zapcore.WriteSyncer
+	syncer = os.Stdout
+	if w != nil {
+		syncer = w
+	}
```
```suggestion
	if w != nil {
		w = os.Stdout
	}
```

yuzhiquan replied:
If w is set, will add a syncer, with w, to logger, that's allow set another output for jsonlogger.

serathius replied:
Yes, so basically `w` default value should be `os.Stdout`
I propose to use `w` instead of creating additional variable `syncer`. That way we reduce number of variables and make it more obvious that this code implements "default pattern".

yuzhiquan replied:
I got this.

yuzhiquan commented:
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-kind-ipv6

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

serathius commented:
I compared this code to https://github.com/go-logr/zapr/blob/master/zapr.go and there are not a lot of differences. 
I think it would make sense to discuss porting those changes to zapr instead of copying code. I think it makes sense as this repo is owned by k8s contributors. 

/cc @thockin 
What do you think about this?

yuzhiquan commented:
> I compared this code to https://github.com/go-logr/zapr/blob/master/zapr.go and there are not a lot of differences.
> I think it would make sense to discuss porting those changes to zapr instead of copying code. I think it makes sense as this repo is owned by k8s contributors.
> 
> /cc @thockin
> What do you think about this?

Sure, it's definitely necessary.

fedebongio commented:
/cc @logicalhan

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

serathius commented:
I'm ok with current implementation. Implementing json is blocking adding flags to components, so let's consider merging code with zapr as next step. I would be good that we start collecting approvals.
/lgtm 
/approve

serathius commented:
/ok-to-test

serathius commented:
/assign @sttts @mtaufen

yuzhiquan commented:
/retest

serathius commented:
Hey @yuzhiquan `pull-kubernetes-verify` is static analysis that is not flaky. No need to retry this one. This is the error that is reported
```
staging/src/k8s.io/component-base/logs/registry.go:	jsonL "k8s.io/component-base/logs/json"
!!! Some package aliases break go conventions.
To fix these errors, do not use capitalized or underlined characters
in pkg aliases. Refer to https://blog.golang.org/package-names for more info.
```
If I understand it correctly, it complains about `jsonL` alias for imports. Can you just remove alias?

yuzhiquan commented:
> Hey @yuzhiquan `pull-kubernetes-verify` is static analysis that is not flaking. No need to retry this one. This is the error that is reported
> 
> ```
> staging/src/k8s.io/component-base/logs/registry.go:	jsonL "k8s.io/component-base/logs/json"
> !!! Some package aliases break go conventions.
> To fix these errors, do not use capitalized or underlined characters
> in pkg aliases. Refer to https://blog.golang.org/package-names for more info.
> ```
> 
> If I understand it correctly, it complains about `jsonL` alias for imports. Can you just remove alias?

Responded

yuzhiquan commented:
Oops, forgot update bazel.,wait a minute.

serathius commented on staging/src/k8s.io/component-base/logs/registry.go:
```diff
@@ -77,5 +81,6 @@ func (lfr *LogFormatRegistry) Delete(name string) {
 	// Text format is default klog format
-	logRegistry.Register("text", nil)
+	logRegistry.Register(defaultLogFormat, nil)
+	logRegistry.Register(jsonLogFormat, json.JSONLogger)
```
I think you forgot to import json

yuzhiquan commented:
/retest

serathius commented:
/lgtm
@liggitt @luxas Can you take a look?

serathius commented:
Hey @yuzhiquan do you think this PR is mergable? if so please remove `[WIP]` prefix from name.

yuzhiquan commented:
Sure, sorry for forget this.

mtaufen commented on staging/src/k8s.io/component-base/logs/json/json.go:
```diff
@@ -0,0 +1,191 @@
+		// check just in case for strongly-typed Zap fields, which is illegal (since
+		// it breaks implementation agnosticism), so we can give a better error message.
+		if _, ok := args[i].(zap.Field); ok {
+			l.DPanic("strongly-typed Zap Field passed to logr", zap.Any("zap field", args[i]))
```
Do we expose an option to log in development vs. production mode?

yuzhiquan replied:
No option  for this now.

serathius replied:
Development mode generates non json format.

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

mtaufen commented on staging/src/k8s.io/component-base/logs/json/json_test.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+// TestInfoLoggerInfo test infologger json info format
+func TestInfoLoggerInfo(t *testing.T) {
```
Can we also test the error conditions have the expected behavior?

E.g. things that cause l.Dpanic, like "strongly typed Zap field" or "non-string key argument"

mtaufen commented:
/approve

serathius commented:
/assign @sttts

sttts commented:
/approve

serathius commented:
@yuzhiquan Please address @mtaufen comments and I will lgtm

sttts commented on staging/src/k8s.io/component-base/logs/registry.go:
```diff
@@ -21,6 +21,11 @@ import (
+)
+
+const (
+	jsonLogFormat = "json"
```
nit: what does this constant add?

yuzhiquan replied:
I think the  json log format string key should be fixed.

yuzhiquan commented:
> @yuzhiquan Please address @mtaufen comments and I will lgtm

Respond.

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

yuzhiquan commented:
Run ./hack/verify-vendor.sh local, nothing more.And run hack/update-vendor.sh local, nothing more too.So i don't known how to fix this build error.

yuzhiquan commented:
/retest

yuzhiquan commented:
/retest

serathius commented:
ping @davidz627 @dims

yuzhiquan commented:
Rebase.
/assign @davidz627 @dims
ping @davidz627 @dims

yuzhiquan commented:
/retest

serathius commented:
pinging other approvers
 @saad-ali @msau42
@andrewsykim

serathius commented:
@lavalamp

dims commented:
/approve
/assign @liggitt

lavalamp commented:
/approve

for go.sum file

serathius commented:
/lgtm

serathius commented:
/retest
